Add multidex flag and automatic multidex support

GaryQian opened the pull request:
This PR adds multi dex support to the flutter_tool. Resolves https://github.com/flutter/flutter/issues/21009

When the multi dex gradle build error is detected, Flutter will prompt the developer whether they wish to enable multi dex support. If yes, android/app/src/main/java/io/flutter/app/FlutterMultiDexApplication.java will be added and the build will be retried.

Once enabled, the existence of the files will be used to determine if the multidex gradle groovy configs and dependencies should added to the app. In practice, this means that apps will build without issue after selecting enable.

Developers who don't want this may pass the --no-multidex flag to disable this automated support for flutter run and flutter build commands.

FlutterMultiDexApplication is codegen-ed rather than included in the engine to reduce engine dependencies and avoid the overhead to non-multidex apps.

GaryQian committed:
Add multidex flag

GaryQian committed:
Add multidex option to run and apk

Clean print

cleanup'

Add error handler

Add keep file gen

Automated

GaryQian committed:
Prompt for enabling multidex

Analyzer

GaryQian commented:
Would prefer if the approach is reviewed first, will add tests shortly.

blasten commented on packages/flutter_tools/gradle/flutter.gradle:
```diff
@@ -149,6 +149,24 @@ class FlutterPlugin implements Plugin<Project> {
+                }
+                buildTypes {
+                    release {
+                        multiDexKeepFile project.file('multidex-flutter-keepfile.txt')
```
could this file be moved to the flutter installation? 

flutter_proguard_rules.pro had a similar issue:
https://github.com/flutter/flutter/blob/e7a9adeeabccf99d67321dc7a87fdf3820f06912/packages/flutter_tools/gradle/flutter.gradle#L188

GaryQian replied:
Yeah!

blasten commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -815,6 +815,16 @@ abstract class FlutterCommand extends Command<void> {
   }
 
+  void addMultiDexOption({ bool hide = false }) {
+    argParser.addFlag('multi-dex',
```
as weird as it sounds, most docs I found refer to "multidex" as a single word.

GaryQian replied:
I was debating to use the one word vs two word version. I think technically, it is two, but I did notice that a lot of docs used the single word, so I will probably just switch to the single word to remain consistent.

blasten commented:
other than the two comments, approach LGTM

GaryQian committed:
Move keepfile into tools

Rename

Fix test

Tests

whitespace

Analyzer

GaryQian committed:
Support flutter drive

Tests

Fix keep file

Better flow for CI builds/run to skip multidex handling

Fix tests

test fix

GaryQian commented:
In testing with lower API devices, turns out multidex is incompatible with reflection. Moving forwards with an alternative solution.

GaryQian commented:
New approach:

- codegen FlutterMultiDexApplication.java, thus avoiding adding overhead to non-multidex apps by not including it as an engine dependency. This also makes it easier for developers to integrate it with any custom android implementations.
- Use manifest variable injection to automatically inject either android.app.Application or io.flutter.app.FlutterMultiDexApplication depending on if multidex is enabled or not. The template should now include a variable for `<application android:name="{$applicationName}>` property in AndroidManifest.xml
- For older apps that do not have the variable injection in their manifest, we ask them to manually add the line.
- Newly generated apps should already have the variable injection and can automatically enable multidex.

GaryQian committed:
New approach

Fix and add tests

Gradle errors tests

Analyzer

GaryQian committed:
Handle V1 vs V2 embedding base application

Analyzer

Typo

Fix base application

Analyzer

GaryQian committed:
Fix android gradle build tests

Check existence of base-application-name

GaryQian committed:
Fix existing tests

blasten reviewed (approved):
This is awesome!

GaryQian committed:
Build tests

GaryQian committed:
Address comments

Address comments

analyzer

Fix tests

Fix tests

Fix tests

GaryQian committed:
Analyzer

GaryQian committed:
address nits

GaryQian committed:
Check specifically for dex failure

GaryQian committed:
Trailing commas

GaryQian committed:
Add logger message
